flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1198,16 +1205,146 @@ class DidNotConverge(Exception):
+        dataset. This is used to verify that writes are persisted to the
+        underlying blockdevice.
+    :ivar _dataset_id: The dataset_id of the dataset under test.
+    :ivar _node_uuid: The node_uuid of the node that is being tested.
+    :ivar _testing_mountroot: A mountroot to construct moutpoints under.
+    :ivar _blockdevice_manager: An ``IBlockDeviceManager`` provider that can be
+        used to mount block devices.
     """
-    def setUp(self):
-        self.deployer = create_blockdevicedeployer(self)
 
-    def teardown_example(self, token):
+    def __init__(self, dataset_id, node_uuid, mountroot, blockdevice_manager):
+        self._known_mountpoints = set()
+        self._device_path = None
+        self._dataset_id = dataset_id
+        self._node_uuid = node_uuid
+        self._testing_mountroot = mountroot
+        self._blockdevice_manager = blockdevice_manager
+
+    def _has_file_named(self, filename):
         """
-        Cleanup after running a hypothesis example.
+        Returns whether the blockdevice at self._device_path has a file with
+        the given name.
+        """
+        test_mountpoint = self._testing_mountroot.child(str(uuid4()))
+        test_mountpoint.makedirs()
+        try:
+            self._blockdevice_manager.mount(self._device_path, test_mountpoint)
+            result = test_mountpoint.child(filename).exists()
+            self._blockdevice_manager.unmount(self._device_path)
+            test_mountpoint.remove()
+        except MountError:
+            # We might not be able to mount if we do not have a filesystem on
+            # the blockdevice.
+            return False
+        return result
+
+    def invariant(self, case, current_cluster_state):
+        """
+        An invariant that should always hold. Specifically, after the path
+        to a dataset is known by the cluster then writes to that path
+        should either succeed and be persisted to the underlying block
+        device, or the write should fail.
+
+        :param case: A ``TestCase`` to use for making assertions.
+        :param current_cluster_state: A callable that returns the current state
+            of the cluster.
+        """
+        # First poll the cluster state to see if there is a new mountpoint that
+        # we should try to use to attempt to write to the dataset.
+        [node] = list(n for n in current_cluster_state().nodes
+                      if n.uuid == self._node_uuid)
+        paths = node.paths or {}
+        path = paths.get(unicode(self._dataset_id))
+        if path:
+            self._known_mountpoints.add(path)
+        devices = node.devices or {}
+        device = devices.get(self._dataset_id)
+        if device:
+            self._device_path = device
+
+        for path in self._known_mountpoints:
+            device = self._device_path
+            filename = str(uuid4())
+            try:
+                path.child(filename).touch()
```

Comment:
(1) Realized it was pretty simple to inspect the OS for a list of all mounts and re-use existing mounts of the backing blockdevice, and then fall back to re-mounting the blockdevice only if we had to. Done.